### PR TITLE
Add test for inhomogeneous term that during numerical integration is equal to zero

### PR DESCRIPTION
**Pull request by clinssen**
If the dynamics of the system is defined such that there is an inhomogeneous (constant-term) component in an ODE, but if numerically, if the inhomogeneous part is zero (this could happen in the case that one or more parameters get set to zero), then because of the way the update expressions are formulated, you could get a division by zero and this results in NaNs in the output.

To address this (but not completely fix it, see #89), ODE-toolbox should automatically check whether setting the inhomogenous part to zero results in a division by zero, and if this is the case, issue a warning. This is akin to checking for singularities in the propagators, but it occurs for the update expressions instead.

A flag is added so that singularity detection (also singularity detection in the propagators) can be disabled for the sake of faster processing. On some systems of equations, singularity detection can take a very long time.

### CODE REVIEW

File: tests/test_singularity_detection.py [hunk 1]
```diff
@@ -32,9 +35,9 @@ class TestSingularityDetection:
     def test_is_matrix_defined_under_substitution(self):
         tau_m, tau_r, C, h = sympy.symbols("tau_m, tau_r, C, h")
         P = sympy.Matrix([[-1 / tau_r, 0, 0], [1, -1 / tau_r, 0], [0, 1 / C, -1 / tau_m]])
-        assert SingularityDetection._is_matrix_defined_under_substitution(P, {})
-        assert SingularityDetection._is_matrix_defined_under_substitution(P, {tau_r: 1})
-        assert not SingularityDetection._is_matrix_defined_under_substitution(P, {tau_r: 0})
+        assert SingularityDetection._is_matrix_defined_under_substitution(P, set())
+        assert SingularityDetection._is_matrix_defined_under_substitution(P, set([sympy.Eq(tau_r, 1)]))
```
Comment (pnbabu) [inline tests/test_singularity_detection.py:39]: Should this be `SymmetricEq`?

File: tests/test_singularity_detection.py [hunk 2]
```diff
@@ -47,9 +50,8 @@ def test_alpha_beta_kernels(self, kernel_to_use: str):
             A = sympy.Matrix([[-1 / tau_d, 0, 0], [1, -1 / tau_r, 0], [0, 1 / C, -1 / tau_m]])
 
         P = sympy.simplify(sympy.exp(A * h))  # Propagator matrix
-
         condition = SingularityDetection._generate_singularity_conditions(P)
-        condition = SingularityDetection._flatten_conditions(condition)  # makes a list of conditions with each condition in the form of a dict
+        print(condition)
```
Comment (pnbabu) [inline tests/test_singularity_detection.py:54]: Could you remove the print statements?

File: tests/test_singularity_detection.py [hunk 3]
```diff
@@ -60,9 +62,41 @@ def test_alpha_beta_kernels(self, kernel_to_use: str):
...
+        log_stream = io.StringIO()
+        log_handler = logging.StreamHandler(log_stream)
+        logger.addHandler(log_handler)
+
+        solver_dict = odetoolbox.analysis(indict, disable_stiffness_check=True, log_level="DEBUG")
+
+        log_contents = log_stream.getvalue()
+
+        # test that no singularity conditions were found for this system
+        assert not "Under certain conditions" in log_contents
```
Comment (pnbabu) [inline tests/test_singularity_detection.py:102]: I think it is better to check for the singularity keyword as well here.

**Review by pnbabu — CHANGES_REQUESTED** [review]

File: tests/test_singularity_detection.py (hunk 1 above)
Comment (clinssen) [reply tests/test_singularity_detection.py:39]: Thanks, fixed!

**Review by pnbabu — APPROVED** [review]
Looks good to me, thanks!
